src/cluster_legacy.c

```diff
@@ -3540,6 +3539,8 @@ static void clusterBuildMessageHdr(clusterMsg *hdr, int type, size_t msglen) {
     /* Set the message flags. */
     if (clusterNodeIsMaster(myself) && server.cluster->mf_end)
         hdr->mflags[0] |= CLUSTERMSG_FLAG0_PAUSED;
+    hdr->mflags[0] |= CLUSTERMSG_FLAG0_EXT_DATA; /* Always make other nodes know that
+                                                  * this node supports extension data. */
```

Comment:
i'm not familiar with this area in the code, maybe you can elaborate.

it looks like before your commit, we set that flag only when the other node doesn't support extensions.
which seems odd. if anything we should set new flags only when talking to new nodes?
although it does make mode sense to always set the flag (assuming old nodes ignore it, rather than breaking on unknown flags).

do you know why Valkey didn't do that too?
did you test the new code against a wide range of versions?

Reply:
In fact, Valkey now always sets this flag, but adds it when writing epingext.
I prefer to build hdr and always set this flag.
```
if (link->node && nodeSupportsExtensions(link->node)) {
    totlen += writePingExt(hdr, gossipcount); <- always set the mark
} else {
    serverLog(LL_DEBUG, "Unable to send extensions data, however setting ext data flag to true");
    hdr->mflags[0] |= CLUSTERMSG_FLAG0_EXT_DATA; <- if node doesn't support, we still set the flag.
}
```

Reply:
> do you know why Valkey didn't do that too? did you test the new code against a wide range of versions?

yes, i tested it between 6.2, 7.0,and 7.2

Reply:
but they still set it conditionally. only when the node doesn't support extensions (or is NULL?)

Reply:
the reason of the issue is that the new version of the node sends ext data to nodes that do not support it.
This causes a failure due to a mismatch between the expected length (explen) and the actual total length of the received data.
https://github.com/redis/redis/blob/731f2dc5c7a6271708a6a79fccf91dda0da74804/src/cluster_legacy.c#L2731-L2736

this PR does the follow things:
1. always set CLUSTERMSG_FLAG0_EXT_DATA, because during the meet phase, we do not support whether the connected node supports ext data, we need to make sure that it knows and brings ext data.
2. if another node does not support ext data, we will not send it ext data to avoid the handshake failure mentioned above due to explen inconsistencies.

Reply:
> but they still set it conditionally. only when the node doesn't support extensions (or is NULL?)

 no, if node support ext data, we always set the flag in the `writePingExt` no matter the extensions is 0 or not.

```
    if (hdr != NULL) {
        hdr->mflags[0] |= CLUSTERMSG_FLAG0_EXT_DATA;
        hdr->extensions = htons(extensions);
    }
```